File: src/M4DSA1/arrays1/add_one_to_number.cpp
#include<iostream>
#include<vector>

using namespace std;

// Time Complexity: O(n) | Space Complexity: O(n)

class Solution {
    public:
    vector<int> plusOne(vector<int> &A) {
        int start = -1;
        vector<int> result;
        bool shouldAdd = true;

        for(int i=0; i<A.size(); i++) {
            if(A[i] != 0) {
                start = i;
                break;
            }
        }

        if(start == -1) return {1};

        for(int i=A.size()-1; i>=start; i--){
            if(shouldAdd) {
                if(A[i] == 9) {
                    result.push_back(0);
                    if(start == i) result.push_back(1);
                    continue;
                }

                result.push_back(A[i] + 1);
                shouldAdd = false;
                continue;
            } else {
                result.push_back(A[i]);
            }
        }

        // swaping 0 with size-1, 1 with size-2...
        for(int i=0; i<result.size()/2; i++) { 
            result[i] += result[result.size()-i-1];
            result[result.size()-i-1] = result[i] - result[result.size()-i-1];
            result[i] -= result[result.size()-i-1];
        }

        return result;
    }
};

int main() {
    vector<int> input = {2, 5, 6, 8, 6, 1, 2, 4, 5};
    Solution solution;
    vector<int> result = solution.plusOne(input);

    for(auto item: result) cout << item << " ";
}

File: src/M4DSA1/arrays1/add_one_to_number.py
class Solution:
    def plus_one(self, arr):
        start, shouldAdd, result = -1, True, []

        for i in range(len(arr)):
            if arr[i] != 0: 
                start = i
                break
        
        if start == -1: return [1]

        for i in range(len(arr) - 1, start - 1, -1):
            if shouldAdd:
                if arr[i] == 9:
                    result.append(0)
                    if start == i: 
                        result.append(i)
                    continue

                result.append(arr[i] + 1)
                shouldAdd = False
                continue
            else:
                result.append(arr[i])

        for i in range(len(arr) // 2):
            result[i], result[len(result) - i - 1] = ( result[len(result) - i - 1], result[i])

        return result
    
solution = Solution()
print(solution.plus_one([2, 5, 6, 8, 6, 1, 2, 4, 5]))